File: packages/skill/src/frontmatter.ts
import {
  MAX_ARGUMENT_HINT_LENGTH,
  MAX_COMPATIBILITY_LENGTH,
  MAX_DESCRIPTION_LENGTH,
  MAX_NAME_LENGTH,
  SKILL_NAME_RE,
} from "./constants.js";

/**
 * Parse result for a SKILL.md file.
 */
export interface ParsedSkill {
  readonly frontmatter: Readonly<Record<string, unknown>>;
  readonly body: string;
}

export interface FrontmatterError {
  readonly kind: "frontmatter_error";
  readonly reason: string;
  readonly line?: number;
}

function isFrontmatterError(x: unknown): x is FrontmatterError {
  return (
    typeof x === "object" &&
    x !== null &&
    (x as { kind?: unknown }).kind === "frontmatter_error"
  );
}

/**
 * Split a SKILL.md into frontmatter text + body. Returns null when there
 * is no frontmatter (no leading `---` line); the whole file is treated
 * as body per agentskills.io.
 *
 * Missing closing `---` after an opening one is an error — we don't
 * silently swallow a malformed header.
 */
export function splitFrontmatter(
  text: string,
): { fmText: string; body: string } | null | FrontmatterError {
  // Normalize CRLF to LF for parsing; preserve original body if no frontmatter.
  const normalized = text.replace(/\r\n/g, "\n");

  // Must start with --- on its own line. Leading BOM is tolerated.
  const stripped = normalized.startsWith("\uFEFF")
    ? normalized.slice(1)
    : normalized;

  if (!stripped.startsWith("---\n") && stripped !== "---") {
    // No frontmatter — whole file is body.
    return null;
  }

  // Find the closing ---. Must be on its own line.
  const lines = stripped.split("\n");
  let close = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i] === "---") {
      close = i;
      break;
    }
  }
  if (close === -1) {
    return {
      kind: "frontmatter_error",
      reason:
        "frontmatter has an opening `---` but no closing `---`; the file must have YAML between two `---` lines at the top",
    };
  }

  const fmText = lines.slice(1, close).join("\n");
  const body = lines.slice(close + 1).join("\n");
  return { fmText, body };
}

/**
 * Minimal YAML parser scoped to SKILL.md frontmatter. Supports:
 *
 * - `key: value` (scalar)
 * - `key: "value"` / `key: 'value'` (quoted scalar)
 * - `key: true|false` (boolean)
 * - `key: 42` (integer)
 * - `key: |` followed by indented block lines (literal block scalar)
 * - `key:` followed by a nested one-level map (indented key: value)
 * - `key: [a, b, c]` (flow-style array of strings)
 * - `key:` followed by `- a`, `- b` lines (block-style array of strings)
 *
 * Anything else is preserved as a raw string via best-effort, and
 * unexpected YAML constructs (anchors, aliases, complex types, nested
 * flows) are rejected with a structured error. We keep it small
 * intentionally — skills converge on the same ~10 fields.
 */
export function parseYamlFrontmatter(
  fmText: string,
): Record<string, unknown> | FrontmatterError {
  const out: Record<string, unknown> = {};
  const lines = fmText.split("\n");
  let i = 0;
  while (i < lines.length) {
    const rawLine = lines[i];
    if (rawLine === undefined) {
      i++;
      continue;
    }
    if (isBlankOrComment(rawLine)) {
      i++;
      continue;
    }
    // Top-level key must not be indented.
    if (startsWithWhitespace(rawLine)) {
      return {
        kind: "frontmatter_error",
        reason: `unexpected indentation at top-level line ${i + 1}`,
        line: i + 1,
      };
    }
    const colonIdx = rawLine.indexOf(":");
    if (colonIdx === -1) {
      return {
        kind: "frontmatter_error",
        reason: `expected 'key: value' on line ${i + 1}`,
        line: i + 1,
      };
    }
    const key = rawLine.slice(0, colonIdx).trim();
    if (!key) {
      return {
        kind: "frontmatter_error",
        reason: `empty key on line ${i + 1}`,
        line: i + 1,
      };
    }
    const rest = rawLine.slice(colonIdx + 1);
    const inline = rest.trim();

    // Case 1: literal block scalar — `key: |`
    if (inline === "|" || inline === "|-" || inline === "|+") {
      const [value, consumed] = readLiteralBlock(lines, i + 1);
      out[key] = value;
      i = consumed;
      continue;
    }

    // Case 2: value present on same line.
    if (inline.length > 0) {
      // Flow-style array: `[a, b, c]`
      if (inline.startsWith("[") && inline.endsWith("]")) {
        out[key] = parseFlowArray(inline);
        i++;
        continue;
      }
      out[key] = parseScalar(inline);
      i++;
      continue;
    }

    // Case 3: value is nested on following indented lines.
    // Look at the next non-blank line to decide: map vs array.
    const nextNonBlank = findNextNonBlank(lines, i + 1);
    if (nextNonBlank === -1) {
      // Nothing follows — empty value.
      out[key] = "";
      i++;
      continue;
    }
    const nextLine = lines[nextNonBlank]!;
    if (!startsWithWhitespace(nextLine)) {
      // The next top-level key — current key has empty value.
      out[key] = "";
      i++;
      continue;
    }
    if (nextLine.trimStart().startsWith("- ")) {
      // Block array of strings.
      const [arr, consumed] = readBlockArray(lines, i + 1);
      out[key] = arr;
      i = consumed;
      continue;
    }
    // Nested map (one level). We only support string → scalar keys.
    const nested = readNestedMap(lines, i + 1);
    if (!Array.isArray(nested)) {
      return nested;
    }
    const [map, consumed] = nested;
    out[key] = map;
    i = consumed;
  }
  return out;
}

function isBlankOrComment(line: string): boolean {
  const t = line.trim();
  return t.length === 0 || t.startsWith("#");
}

function startsWithWhitespace(line: string): boolean {
  return /^[ \t]/.test(line);
}

function findNextNonBlank(lines: readonly string[], from: number): number {
  for (let i = from; i < lines.length; i++) {
    const line = lines[i];
    if (line === undefined) continue;
    if (!isBlankOrComment(line)) return i;
  }
  return -1;
}

function parseScalar(s: string): unknown {
  if (
    (s.startsWith('"') && s.endsWith('"')) ||
    (s.startsWith("'") && s.endsWith("'"))
  ) {
    return s.slice(1, -1);
  }
  if (s === "true") return true;
  if (s === "false") return false;
  if (s === "null" || s === "~") return null;
  if (/^-?\d+$/.test(s)) return Number.parseInt(s, 10);
  if (/^-?\d+\.\d+$/.test(s)) return Number.parseFloat(s);
  return s;
}

function parseFlowArray(s: string): string[] {
  const inner = s.slice(1, -1).trim();
  if (!inner) return [];
  return inner.split(",").map((item) => {
    const t = item.trim();
    if (
      (t.startsWith('"') && t.endsWith('"')) ||
      (t.startsWith("'") && t.endsWith("'"))
    ) {
      return t.slice(1, -1);
    }
    return t;
  });
}

function readLiteralBlock(
  lines: readonly string[],
  from: number,
): [string, number] {
  // Find indentation of first non-blank block line.
  let first = from;
  while (first < lines.length) {
    const line = lines[first];
    if (line === undefined) break;
    if (line.trim().length === 0) {
      first++;
      continue;
    }
    break;
  }
  if (first >= lines.length) return ["", from];
  const firstLine = lines[first]!;
  const indent = firstLine.match(/^[ \t]+/)?.[0] ?? "";
  if (!indent) return ["", from];
  const collected: string[] = [];
  let i = from;
  while (i < lines.length) {
    const line = lines[i];
    if (line === undefined) break;
    if (line.trim().length === 0) {
      collected.push("");
      i++;
      continue;
    }
    if (!line.startsWith(indent)) break;
    collected.push(line.slice(indent.length));
    i++;
  }
  // Trim trailing blanks conservatively (spec equivalent of `|-`).
  while (collected.length > 0 && collected[collected.length - 1] === "") {
    collected.pop();
  }
  return [collected.join("\n"), i];
}

function readBlockArray(
  lines: readonly string[],
  from: number,
): [string[], number] {
  const arr: string[] = [];
  let i = from;
  let indent: string | null = null;
  while (i < lines.length) {
    const line = lines[i];
    if (line === undefined) break;
    if (isBlankOrComment(line)) {
      i++;
      continue;
    }
    const leading = line.match(/^[ \t]+/)?.[0] ?? "";
    if (!leading) break;
    if (indent === null) indent = leading;
    if (leading !== indent) break;
    const after = line.slice(leading.length);
    if (!after.startsWith("- ") && after !== "-") break;
    const value = after === "-" ? "" : after.slice(2).trim();
    arr.push(typeof parseScalar(value) === "string" ? (parseScalar(value) as string) : String(parseScalar(value)));
    i++;
  }
  return [arr, i];
}

function readNestedMap(
  lines: readonly string[],
  from: number,
): [Record<string, unknown>, number] | FrontmatterError {
  const map: Record<string, unknown> = {};
  let i = from;
  let indent: string | null = null;
  while (i < lines.length) {
    const line = lines[i];
    if (line === undefined) break;
    if (isBlankOrComment(line)) {
      i++;
      continue;
    }
    const leading = line.match(/^[ \t]+/)?.[0] ?? "";
    if (!leading) break;
    if (indent === null) indent = leading;
    if (!line.startsWith(indent)) break;
    if (leading.length < indent.length) break;
    const content = line.slice(indent.length);
    const colonIdx = content.indexOf(":");
    if (colonIdx === -1) {
      return {
        kind: "frontmatter_error",
        reason: `expected 'key: value' in nested map on line ${i + 1}`,
        line: i + 1,
      };
    }
    const key = content.slice(0, colonIdx).trim();
    const rest = content.slice(colonIdx + 1).trim();
    if (rest.startsWith("[") && rest.endsWith("]")) {
      map[key] = parseFlowArray(rest);
    } else {
      map[key] = parseScalar(rest);
    }
    i++;
  }
  return [map, i];
}

/**
 * Full validation: parse + required fields + constraints + name match.
 */
export interface ValidateArgs {
  readonly fmText: string;
  readonly body: string;
  /** Basename of the containing skill dir; must equal frontmatter.name. */
  readonly expectedName: string;
}

export interface ValidatedSkill {
  readonly kind: "ok";
  readonly frontmatter: Readonly<Record<string, unknown>>;
  readonly body: string;
}

export interface ValidationError {
  readonly kind: "error";
  readonly reason: string;
  readonly code: "INVALID_FRONTMATTER" | "NAME_MISMATCH";
  readonly line?: number;
}

export function validateFrontmatter(
  args: ValidateArgs,
): ValidatedSkill | ValidationError {
  const parsed = parseYamlFrontmatter(args.fmText);
  if (isFrontmatterError(parsed)) {
    return {
      kind: "error",
      code: "INVALID_FRONTMATTER",
      reason: parsed.reason,
      ...(parsed.line !== undefined ? { line: parsed.line } : {}),
    };
  }
  const fm: Record<string, unknown> = parsed;

  // Required: name
  const name = fm.name;
  if (typeof name !== "string" || name.length === 0) {
    return {
      kind: "error",
      code: "INVALID_FRONTMATTER",
      reason: "frontmatter missing required field 'name'",
    };
  }
  if (name.length > MAX_NAME_LENGTH) {
    return {
      kind: "error",
      code: "INVALID_FRONTMATTER",
      reason: `frontmatter 'name' exceeds ${MAX_NAME_LENGTH} chars`,
    };
  }
  if (!SKILL_NAME_RE.test(name)) {
    return {
      kind: "error",
      code: "INVALID_FRONTMATTER",
      reason: `frontmatter 'name' must match lowercase-kebab-case regex ${SKILL_NAME_RE.source}; got "${name}"`,
    };
  }

  // Required: description
  const description = fm.description;
  if (typeof description !== "string" || description.length === 0) {
    return {
      kind: "error",
      code: "INVALID_FRONTMATTER",
      reason: "frontmatter missing required field 'description'",
    };
  }
  if (description.length > MAX_DESCRIPTION_LENGTH) {
    return {
      kind: "error",
      code: "INVALID_FRONTMATTER",
      reason: `frontmatter 'description' exceeds ${MAX_DESCRIPTION_LENGTH} chars`,
    };
  }

  // Optional constraints
  const compat = fm.compatibility;
  if (compat !== undefined) {
    if (typeof compat !== "string") {
      return {
        kind: "error",
        code: "INVALID_FRONTMATTER",
        reason: "'compatibility' must be a string",
      };
    }
    if (compat.length > MAX_COMPATIBILITY_LENGTH) {
      return {
        kind: "error",
        code: "INVALID_FRONTMATTER",
        reason: `'compatibility' exceeds ${MAX_COMPATIBILITY_LENGTH} chars`,
      };
    }
  }

  const hint = fm["argument-hint"];
  if (hint !== undefined) {
    if (typeof hint !== "string") {
      return {
        kind: "error",
        code: "INVALID_FRONTMATTER",
        reason: "'argument-hint' must be a string",
      };
    }
    if (hint.length > MAX_ARGUMENT_HINT_LENGTH) {
      return {
        kind: "error",
        code: "INVALID_FRONTMATTER",
        reason: `'argument-hint' exceeds ${MAX_ARGUMENT_HINT_LENGTH} chars`,
      };
    }
  }

  // allowed-tools can be a string or a string[].
  if (fm["allowed-tools"] !== undefined) {
    const at = fm["allowed-tools"];
    if (typeof at === "string") {
      // Normalize to array.
      const items = at
        .split(/[,\s]+/)
        .map((s) => s.trim())
        .filter((s) => s.length > 0);
      fm["allowed-tools"] = items;
    } else if (!Array.isArray(at)) {
      return {
        kind: "error",
        code: "INVALID_FRONTMATTER",
        reason: "'allowed-tools' must be a string or string[]",
      };
    }
  }

  // Name must match containing directory.
  if (name !== args.expectedName) {
    return {
      kind: "error",
      code: "NAME_MISMATCH",
      reason: `frontmatter 'name' ("${name}") does not match the skill directory ("${args.expectedName}")`,
    };
  }

  return {
    kind: "ok",
    frontmatter: fm,
    body: args.body,
  };
}

File: crates/skill/src/frontmatter.rs
//! Minimal YAML parser scoped to SKILL.md frontmatter. Matches the TS
//! parser at `packages/skill/src/frontmatter.ts` verbatim in behavior.
//!
//! Supports:
//! - `key: value` (scalar)
//! - `key: "value"` / `key: 'value'` (quoted scalar)
//! - `key: true|false` / `key: 42` / `key: null|~`
//! - `key: |` followed by indented block lines (literal block scalar)
//! - `key:` followed by a nested one-level map (indented `key: value`)
//! - `key: [a, b, c]` (flow-style array of strings)
//! - `key:` followed by `- a`, `- b` block array lines
//!
//! Anything else is rejected with a structured error. We keep it small
//! on purpose — skills converge on the same ~10 fields.

use serde_json::{Map, Value};

use crate::constants::{
    MAX_ARGUMENT_HINT_LENGTH, MAX_COMPATIBILITY_LENGTH, MAX_DESCRIPTION_LENGTH, MAX_NAME_LENGTH,
};
use crate::schema::is_valid_skill_name;

#[derive(Debug, Clone)]
pub struct FrontmatterError {
    pub reason: String,
    pub line: Option<usize>,
}

/// Splits SKILL.md into `(fm_text, body)`. Returns `Ok(None)` if the
/// file has no frontmatter (no leading `---`). Returns `Err` only when
/// there's a malformed opening without a closing `---`.
pub fn split_frontmatter(
    text: &str,
) -> Result<Option<(String, String)>, FrontmatterError> {
    // Normalize CRLF → LF for parsing. Tolerate BOM.
    let normalized = text.replace("\r\n", "\n");
    let stripped = normalized.strip_prefix('\u{FEFF}').unwrap_or(&normalized);

    if !stripped.starts_with("---\n") && stripped != "---" {
        return Ok(None);
    }

    let lines: Vec<&str> = stripped.split('\n').collect();
    let mut close: Option<usize> = None;
    for (i, &line) in lines.iter().enumerate().skip(1) {
        if line == "---" {
            close = Some(i);
            break;
        }
    }
    let close = close.ok_or_else(|| FrontmatterError {
        reason:
            "frontmatter has an opening `---` but no closing `---`; the file must have YAML between two `---` lines at the top"
                .into(),
        line: None,
    })?;

    let fm_text = lines[1..close].join("\n");
    let body = if close + 1 < lines.len() {
        lines[close + 1..].join("\n")
    } else {
        String::new()
    };
    Ok(Some((fm_text, body)))
}

pub fn parse_yaml_frontmatter(
    fm_text: &str,
) -> Result<Map<String, Value>, FrontmatterError> {
    let mut out = Map::new();
    let lines: Vec<&str> = fm_text.split('\n').collect();
    let mut i = 0usize;
    while i < lines.len() {
        let raw = lines[i];
        if is_blank_or_comment(raw) {
            i += 1;
            continue;
        }
        if starts_with_whitespace(raw) {
            return Err(FrontmatterError {
                reason: format!("unexpected indentation at top-level line {}", i + 1),
                line: Some(i + 1),
            });
        }
        let colon_idx = match raw.find(':') {
            Some(c) => c,
            None => {
                return Err(FrontmatterError {
                    reason: format!("expected 'key: value' on line {}", i + 1),
                    line: Some(i + 1),
                });
            }
        };
        let key = raw[..colon_idx].trim();
        if key.is_empty() {
            return Err(FrontmatterError {
                reason: format!("empty key on line {}", i + 1),
                line: Some(i + 1),
            });
        }
        let rest = &raw[colon_idx + 1..];
        let inline = rest.trim();

        // Literal block scalar `key: |`, `key: |-`, `key: |+`.
        if inline == "|" || inline == "|-" || inline == "|+" {
            let (value, consumed) = read_literal_block(&lines, i + 1);
            out.insert(key.to_string(), Value::String(value));
            i = consumed;
            continue;
        }

        // Inline value.
        if !inline.is_empty() {
            if inline.starts_with('[') && inline.ends_with(']') {
                out.insert(key.to_string(), parse_flow_array(inline));
                i += 1;
                continue;
            }
            out.insert(key.to_string(), parse_scalar(inline));
            i += 1;
            continue;
        }

        // Nested block: map or block-array.
        let next_non_blank = find_next_non_blank(&lines, i + 1);
        let Some(nnb) = next_non_blank else {
            out.insert(key.to_string(), Value::String(String::new()));
            i += 1;
            continue;
        };
        let next_line = lines[nnb];
        if !starts_with_whitespace(next_line) {
            out.insert(key.to_string(), Value::String(String::new()));
            i += 1;
            continue;
        }
        if next_line.trim_start().starts_with("- ") || next_line.trim_start() == "-" {
            let (arr, consumed) = read_block_array(&lines, i + 1);
            out.insert(key.to_string(), Value::Array(arr));
            i = consumed;
            continue;
        }
        let (map, consumed) = read_nested_map(&lines, i + 1)?;
        out.insert(key.to_string(), Value::Object(map));
        i = consumed;
    }
    Ok(out)
}

fn is_blank_or_comment(line: &str) -> bool {
    let t = line.trim();
    t.is_empty() || t.starts_with('#')
}

fn starts_with_whitespace(line: &str) -> bool {
    line.starts_with(' ') || line.starts_with('\t')
}

fn find_next_non_blank(lines: &[&str], from: usize) -> Option<usize> {
    (from..lines.len()).find(|&i| !is_blank_or_comment(lines[i]))
}

fn parse_scalar(s: &str) -> Value {
    if (s.starts_with('"') && s.ends_with('"') && s.len() >= 2)
        || (s.starts_with('\'') && s.ends_with('\'') && s.len() >= 2)
    {
        return Value::String(s[1..s.len() - 1].to_string());
    }
    match s {
        "true" => return Value::Bool(true),
        "false" => return Value::Bool(false),
        "null" | "~" => return Value::Null,
        _ => {}
    }
    if let Ok(i) = s.parse::<i64>() {
        return Value::Number(i.into());
    }
    if let Ok(f) = s.parse::<f64>() {
        if let Some(n) = serde_json::Number::from_f64(f) {
            return Value::Number(n);
        }
    }
    Value::String(s.to_string())
}

fn parse_flow_array(s: &str) -> Value {
    let inner = s[1..s.len() - 1].trim();
    if inner.is_empty() {
        return Value::Array(Vec::new());
    }
    let items: Vec<Value> = inner
        .split(',')
        .map(|item| {
            let t = item.trim();
            if (t.starts_with('"') && t.ends_with('"') && t.len() >= 2)
                || (t.starts_with('\'') && t.ends_with('\'') && t.len() >= 2)
            {
                Value::String(t[1..t.len() - 1].to_string())
            } else {
                Value::String(t.to_string())
            }
        })
        .collect();
    Value::Array(items)
}

fn read_literal_block(lines: &[&str], from: usize) -> (String, usize) {
    // Find indentation of first non-blank block line.
    let mut first = from;
    while first < lines.len() {
        if lines[first].trim().is_empty() {
            first += 1;
            continue;
        }
        break;
    }
    if first >= lines.len() {
        return (String::new(), from);
    }
    let first_line = lines[first];
    // Measure leading whitespace.
    let indent: String = first_line
        .chars()
        .take_while(|c| *c == ' ' || *c == '\t')
        .collect();
    if indent.is_empty() {
        return (String::new(), from);
    }
    let mut collected: Vec<String> = Vec::new();
    let mut i = from;
    while i < lines.len() {
        let line = lines[i];
        if line.trim().is_empty() {
            collected.push(String::new());
            i += 1;
            continue;
        }
        if !line.starts_with(&indent) {
            break;
        }
        collected.push(line[indent.len()..].to_string());
        i += 1;
    }
    // Trim trailing blanks conservatively (spec equivalent of `|-`).
    while collected.last().map(|s| s.is_empty()).unwrap_or(false) {
        collected.pop();
    }
    (collected.join("\n"), i)
}

fn read_block_array(lines: &[&str], from: usize) -> (Vec<Value>, usize) {
    let mut arr: Vec<Value> = Vec::new();
    let mut i = from;
    let mut indent: Option<String> = None;
    while i < lines.len() {
        let line = lines[i];
        if is_blank_or_comment(line) {
            i += 1;
            continue;
        }
        let leading: String = line
            .chars()
            .take_while(|c| *c == ' ' || *c == '\t')
            .collect();
        if leading.is_empty() {
            break;
        }
        if indent.is_none() {
            indent = Some(leading.clone());
        }
        if leading != *indent.as_ref().unwrap() {
            break;
        }
        let after = &line[leading.len()..];
        if !after.starts_with("- ") && after != "-" {
            break;
        }
        let value_str = if after == "-" { "" } else { after[2..].trim() };
        arr.push(parse_scalar(value_str));
        i += 1;
    }
    (arr, i)
}

fn read_nested_map(
    lines: &[&str],
    from: usize,
) -> Result<(Map<String, Value>, usize), FrontmatterError> {
    let mut map = Map::new();
    let mut i = from;
    let mut indent: Option<String> = None;
    while i < lines.len() {
        let line = lines[i];
        if is_blank_or_comment(line) {
            i += 1;
            continue;
        }
        let leading: String = line
            .chars()
            .take_while(|c| *c == ' ' || *c == '\t')
            .collect();
        if leading.is_empty() {
            break;
        }
        if indent.is_none() {
            indent = Some(leading.clone());
        }
        let ind = indent.as_ref().unwrap();
        if !line.starts_with(ind) {
            break;
        }
        if leading.len() < ind.len() {
            break;
        }
        let content = &line[ind.len()..];
        let colon_idx = match content.find(':') {
            Some(c) => c,
            None => {
                return Err(FrontmatterError {
                    reason: format!("expected 'key: value' in nested map on line {}", i + 1),
                    line: Some(i + 1),
                });
            }
        };
        let key = content[..colon_idx].trim().to_string();
        let rest = content[colon_idx + 1..].trim();
        if rest.starts_with('[') && rest.ends_with(']') {
            map.insert(key, parse_flow_array(rest));
        } else if rest.starts_with('{') && rest.ends_with('}') {
            // Flow-style inline map like `{type: string}` — parse into an
            // object of string→scalar. Lightweight since only `arguments`
            // uses this shape.
            let inner = rest[1..rest.len() - 1].trim();
            let mut flow_map = Map::new();
            if !inner.is_empty() {
                for pair in inner.split(',') {
                    let pair = pair.trim();
                    if let Some(ci) = pair.find(':') {
                        let k = pair[..ci].trim().to_string();
                        let v = pair[ci + 1..].trim();
                        flow_map.insert(k, parse_scalar(v));
                    }
                }
            }
            map.insert(key, Value::Object(flow_map));
        } else {
            map.insert(key, parse_scalar(rest));
        }
        i += 1;
    }
    Ok((map, i))
}

// ---- validate ----

#[derive(Debug, Clone)]
pub enum ValidationError {
    InvalidFrontmatter { reason: String, line: Option<usize> },
    NameMismatch { reason: String },
}

pub struct ValidatedSkill {
    pub frontmatter: Map<String, Value>,
    pub body: String,
}

pub fn validate_frontmatter(
    fm_text: &str,
    body: &str,
    expected_name: &str,
) -> Result<ValidatedSkill, ValidationError> {
    let mut fm = parse_yaml_frontmatter(fm_text).map_err(|e| {
        ValidationError::InvalidFrontmatter {
            reason: e.reason,
            line: e.line,
        }
    })?;

    // name
    let name = match fm.get("name") {
        Some(Value::String(s)) if !s.is_empty() => s.clone(),
        _ => {
            return Err(ValidationError::InvalidFrontmatter {
                reason: "frontmatter missing required field 'name'".into(),
                line: None,
            });
        }
    };
    if name.len() > MAX_NAME_LENGTH {
        return Err(ValidationError::InvalidFrontmatter {
            reason: format!("frontmatter 'name' exceeds {} chars", MAX_NAME_LENGTH),
            line: None,
        });
    }
    if !is_valid_skill_name(&name) {
        return Err(ValidationError::InvalidFrontmatter {
            reason: format!(
                "frontmatter 'name' must match lowercase-kebab-case regex; got \"{}\"",
                name
            ),
            line: None,
        });
    }

    // description
    let description = match fm.get("description") {
        Some(Value::String(s)) if !s.is_empty() => s.clone(),
        _ => {
            return Err(ValidationError::InvalidFrontmatter {
                reason: "frontmatter missing required field 'description'".into(),
                line: None,
            });
        }
    };
    if description.len() > MAX_DESCRIPTION_LENGTH {
        return Err(ValidationError::InvalidFrontmatter {
            reason: format!(
                "frontmatter 'description' exceeds {} chars",
                MAX_DESCRIPTION_LENGTH
            ),
            line: None,
        });
    }

    // compatibility
    if let Some(v) = fm.get("compatibility") {
        match v {
            Value::String(s) => {
                if s.len() > MAX_COMPATIBILITY_LENGTH {
                    return Err(ValidationError::InvalidFrontmatter {
                        reason: format!(
                            "'compatibility' exceeds {} chars",
                            MAX_COMPATIBILITY_LENGTH
                        ),
                        line: None,
                    });
                }
            }
            _ => {
                return Err(ValidationError::InvalidFrontmatter {
                    reason: "'compatibility' must be a string".into(),
                    line: None,
                });
            }
        }
    }

    // argument-hint
    if let Some(v) = fm.get("argument-hint") {
        match v {
            Value::String(s) => {
                if s.len() > MAX_ARGUMENT_HINT_LENGTH {
                    return Err(ValidationError::InvalidFrontmatter {
                        reason: format!(
                            "'argument-hint' exceeds {} chars",
                            MAX_ARGUMENT_HINT_LENGTH
                        ),
                        line: None,
                    });
                }
            }
            _ => {
                return Err(ValidationError::InvalidFrontmatter {
                    reason: "'argument-hint' must be a string".into(),
                    line: None,
                });
            }
        }
    }

    // allowed-tools: normalize string → array
    if let Some(v) = fm.get("allowed-tools").cloned() {
        match v {
            Value::String(s) => {
                let items: Vec<Value> = s
                    .split(|c: char| c == ',' || c.is_ascii_whitespace())
                    .filter(|t| !t.is_empty())
                    .map(|t| Value::String(t.to_string()))
                    .collect();
                fm.insert("allowed-tools".to_string(), Value::Array(items));
            }
            Value::Array(_) => { /* ok */ }
            _ => {
                return Err(ValidationError::InvalidFrontmatter {
                    reason: "'allowed-tools' must be a string or string[]".into(),
                    line: None,
                });
            }
        }
    }

    // name match against containing dir
    if name != expected_name {
        return Err(ValidationError::NameMismatch {
            reason: format!(
                "frontmatter 'name' (\"{}\") does not match the skill directory (\"{}\")",
                name, expected_name
            ),
        });
    }

    Ok(ValidatedSkill {
        frontmatter: fm,
        body: body.to_string(),
    })
}
